source/extensions/io_socket/buffered_io_socket/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,333 @@
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::read(Buffer::Instance& buffer,
+                                                         uint64_t max_length) {
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  if (pending_received_data_.length() == 0) {
+    if (receive_data_end_stream_) {
+      return {0, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+    } else {
+      return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                                 Network::IoSocketError::deleteIoError)};
+    }
+  }
+  // TODO(lambdai): Move at slice boundary to move to reduce the copy.
+  uint64_t max_bytes_to_read = std::min(max_length, pending_received_data_.length());
+  buffer.move(pending_received_data_, max_bytes_to_read);
+  return {max_bytes_to_read, Api::IoErrorPtr(nullptr, [](Api::IoError*) {})};
+}
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::writev(const Buffer::RawSlice* slices,
+                                                           uint64_t num_slice) {
+  // Empty input is allowed even though the peer is shutdown.
+  bool is_input_empty = true;
+  for (uint64_t i = 0; i < num_slice; i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      is_input_empty = false;
+      break;
+    }
+  }
+  if (is_input_empty) {
+    return Api::ioCallUint64ResultNoError();
+  };
+  if (!isOpen()) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Closed peer.
+  if (!writable_peer_) {
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // Error: write after close.
+  if (writable_peer_->isPeerShutDownWrite()) {
+    // TODO(lambdai): `EPIPE` or `ENOTCONN`.
+    return {0, Api::IoErrorPtr(new Network::IoSocketError(SOCKET_ERROR_INVAL),
+                               Network::IoSocketError::deleteIoError)};
+  }
+  // The peer is valid but temporary not accepts new data. Likely due to flow control.
+  if (!writable_peer_->isWritable()) {
+    return {0, Api::IoErrorPtr(Network::IoSocketError::getIoSocketEagainInstance(),
+                               Network::IoSocketError::deleteIoError)};
+  }
+
+  auto* const dest_buffer = writable_peer_->getWriteBuffer();
+  // Write along with iteration. Buffer guarantee the fragment is always append-able.
+  uint64_t bytes_written = 0;
+  for (uint64_t i = 0; i < num_slice && !dest_buffer->highWatermarkTriggered(); i++) {
+    if (slices[i].mem_ != nullptr && slices[i].len_ != 0) {
+      dest_buffer->add(slices[i].mem_, slices[i].len_);
+      bytes_written += slices[i].len_;
+    }
+  }
+  writable_peer_->setNewDataAvailable();
```

Comment:
This call should be omitted if the buffer was not empty at the beginning of this method.  We're looking for edge trigger behavior, no need to notify of new data if there was data in the buffer prior to the writev call.